src/WebJobs.Script.WebHost/Security/SecretManager.cs

```diff
@@ -446,5 +425,76 @@ private void OnSecretsChanged(object sender, SecretsChangedEventArgs e)
 
             await _repository.PurgeOldSecretsAsync(currentFunctions, traceWriter, logger);
         }
+
+        private async Task<HostSecrets> LoadHostSecretsWithRetryAsync()
+        {
+            int currentRetry = 0;
+            for (; ;)
+            {
+                try
+                {
+                    HostSecrets hostSecrets = await LoadSecretsAsync<HostSecrets>();
+                    if (hostSecrets == null)
+                    {
+                        _traceWriter.Verbose(Resources.TraceHostSecretGeneration);
+                        _logger?.LogDebug(Resources.TraceHostSecretGeneration);
+                        hostSecrets = GenerateHostSecrets();
+                        await PersistSecretsAsync(hostSecrets);
+                    }
+                    return hostSecrets;
+                }
+                catch (Exception ex)
+                {
+                    currentRetry++;
+                    if (currentRetry > _retryCount || !IsTransient(ex))
+                    {
+                        throw;
+                    }
+                }
+                await Task.Delay(_retryDelay);
+            }
+        }
+
+        private async Task<FunctionSecrets> LoadFunctionSecretsWithRetryAsync(string functionName)
```

Comment:
Can we avoid the duplication here by factoring the common fault handling code into a method that takes a delegate?